lisa/energy_model.py

```diff
@@ -626,12 +626,19 @@ def _estimate_from_active_time(self, cpu_active_time, freqs, idle_states,
             _idle_power = max(node.idle_states[idle_states[c]] for c in cpus)
             idle_power = _idle_power * (1 - active_time)
 
+            if cpus not in ret:
+                if combine:
+                    ret[cpus] = 0
+                else:
+                    ret[cpus] = {}
+                    ret[cpus]["active"] = 0
+                    ret[cpus]["idle"] = 0
+
             if combine:
-                ret[cpus] = active_power + idle_power
+                ret[cpus] += active_power + idle_power
```

Comment:
Is the returned dict supposed to have a different key for each CPU and for each cluster ? If so, this will basically assume cluster energy = 0, since we add all the cost to `(n,)`, which is recognized as a CPU by the rest of the code. Is there is anything CPU-specific somewhere, i.e. that needs to make the difference between `(n,)-the-cluster` and `(n,)-the-cpu` ?

Reply:
I don't think anyone in the code path needs the distinction between `(n,)-the-cluster` and `(n,)-the-cpu` estimate_from_cpu_util(). The callers are:

**1. get_optimal_placements():**

```
power = self.estimate_from_cpu_util(util, freqs=freqs)
candidates[util] = sum(power.values())
```

This shouldn't be a problem here.

**2. tests/test_energy_model.py**

The topology used doesn't have a single CPU cluster. The test still passes.

**3. lisa/tests/staging/schedutil.py  estimate_nrg()**

Unused.